### ISSUE TITLE
Relationship with learnbase and JuliaML

### ISSUE THREAD

**Issue by datnamer** [alan-turing-institute/MLJ.jl#47]
Very exciting to learn about this effort! A julia native ML package improving on Sklearn is one of the key missing pieces of the ecosystem.

Here's a list of things I'd like to bring to your attention, if you haven't considered them already. I can open issues for any that deserve their own.

1. There's already been a fairly large effort for something along these lines in the JuliaML ecosystem. Might want to consider integration or lifting of ideas: https://github.com/JuliaML 

2. Integration with prob programming framework like turing.jl would be really cool. They have non gradient samplers that can work with arbitrary julia code, along with HMC. Would be cool if point parameters and priors can be mixed in a model (or model search as prob program induction)  with different sampling/optimization strategies. https://github.com/TuringLang/Turing.jl  cc @yebai

3. Regarding architecture search and "automl", here are some python exemplars: https://github.com/automl/auto-sklearn, https://github.com/jhfjhfj1/autokeras,  https://github.com/EpistasisLab/tpot. tpot can optimize over non differentiable pipelines using genetic programming. 

4. Tables.jl is an alternative table interface which much of the stats ecosystem is coordinating around by @quinnj which can also hook into interable tables, though IIRC there are some issues with missing data interop. 

5. One of those with integration is [statsmodels.jl](https://github.com/JuliaStats/StatsModels.jl), which has a very powerful formula interface that works with abstract tables. Would be cool (and an improvement over sklearn) to integrate with this.

**Comment by datnamer** [comment]
Thanks for your reply. That all makes sense.

Re ppl, it would be nice to be able to wrap a turing.jl model for use in a pipeline without having to expose learmable parameters. Pymc3.has a scikitlearm wrapper to do that.

Also regarding Julia ml, ot was initially meant as a framework for complex ml pipelines with rich abstractions. Here is some info to that effect : https://github.com/JuliaML/LearnBase.jl/blob/master/src/LearnBase.jl

And @evizero.can say more

**Comment by datnamer** [comment]
Here's the right link https://github.com/JuliaML/Transformations.jl

**Comment by fkiraly** [comment]
I think I more or less agree with @tlienart , some other comments:

2. probabilistic programming is essentially the modern, object oriented formulation of Bayesian MCMC (sometimes variational too). As such, it falls into the category of a *method class*, within the Bayesian framework. Bayesians are, in contemporary practice, not very clear on the separation of fitting versus model application (e.g., prediction) vs evaluation that is reflected in sklear's fit/predict design, while the PP world is extremely advanced in modular/abstract model building.

Where the task is prediction, the two can be made to fit together, see e.g., the discussion in section 1.3.4 of
https://arxiv.org/abs/1801.00753
for more details on my opinion on the issue, though it is not immediate as the sklearn an common Bayesian modelling interface designs are differently focused (on model class specific model building). I'd see
https://github.com/alan-turing-institute/skpro
as a design study in how an interface could look like, it is conditional on a probabilistic learning interface which sklearn does *not* have but MLJ now does :-)

I agree with @tlienart that this is an exciting research area (with direct real world applications, e.g., finance or medicine) - though I disagree with the conclusion, I think that makes it especially worthwhile to work on! Though maybe not as one of the current MLJ priorities for pragmatic reasons.

5. We've discussed abstract model specification and model description in #22 - I think the high-tech solution would synergize really well with a formula specification interface such as in statsmodels or R. However, again it might not be a priority given limited resources...

Generally, we're always happy for suggestions/designs/contributions!

The Turing is also always looking for competent people, please consider formally applying here:
https://www.turing.ac.uk/work-turing/research-associates
https://www.turing.ac.uk/work-turing/research-assistants
(part time arrangements are possible)

**Comment by fkiraly** [comment]
Renamed the topic to be more descriptive of its content.

**Comment by ablaom** [comment]
Many thanks @datnamer for your comments and enthusiasm! Just a few things to add to the other comprehensive responses:

3. MLJ already has a flexible API for building learning networks beyond simple linear pipelines, for exporting them as stand-alone models, and tuning their nested parameters. While "architecture" search should be possible, the immediate priority would to improve usability of the existing interface, for example, by providing standard architectures (linear pipelines, stacks, etc) out-of-the-box, and to add to the existing tuning strategies (to include AD gradient descent for pure julia models). 

4. MLJ is indeed attempting to be "data agnostic" and there are two generic tabular data interfaces we have looked at: the Tables.jl interface you refer to, and the Query.jl iterable tablets interface (defined in TableTraits.jl). @ayush1999 and I have played around with these but it is still not absolutely clear to me which is this best. At the moment we are using iterable tables, although we currently have a small intermediate interface in MLJBase that could allow us to change our minds later. An important requirement is integration with CategoricalArrays.jl; some other requirements have been discussed [here](https://github.com/alan-turing-institute/MLJ.jl/issues/15#issuecomment-442520476), another [here](https://discourse.julialang.org/t/query-jl-how-can-i-write-type-preserving-methods-for-iterable-tables/19956). What we have now works but could be improved. 

5. Will have another look at Statsmodels.jl

There are only a few models for which the MLJ interface has been implemented, and a priority is to implement existing the MLJ framework for new models. Any help in this area is particularly appreciated.

**Comment by quinnj** [comment]
On point 4., note that the Tables.jl interface is a superset of the iterable tables set of sources/sinks. I.e. any iterable table is also a Tables.jl-compatible source. This change was made to help simplify the ecosystem and allow for package developers (and use-cases exactly like this) to only need to rely on Tables.jl and get everything else for free. Happy to help answer any other questions regarding using Tables.jl.

**Comment by datnamer** [comment]
Thanks for the feedback everyone. The project seems quite well thought out. Perhaps a blogpost will help garner cooperation from the broader Julia community. 

Glad to know many of these things are or have been considered. 

@fkiraly , regarding PPL integration, I did mean primarily  for prediction and in that vein the reference you linked seems quite intriguing. 

@ablaom When you reconsider statsmodels, please do note that a [monster PR](https://github.com/JuliaStats/StatsModels.jl/pull/71)  by @kleinschmidt is pending which will provide far more  expressiveness and composability to the modeling interface.

**Comment by ablaom** [comment]
@quinn Thanks indeed for those details. After taking another look, I have switched MLJ from Query to Tables, which is serving our needs well for now.

**Comment by DrChainsaw** [comment]
A little late to the party, but here is a related effort w.r.t point 3: https://github.com/DrChainsaw/NaiveGAflux.jl

I'm currently just using it to play around with very loosely restricted search spaces so there is perhaps not much to offer yet in terms of simplify tuning as you'll most likely end up tuning the tuning parameters :) 

If someone has a favourite search space and think it would be useful to add an MLJTuning plugin for it I could probably make an effort to make a package as long as it doesn't require me to shift through python code of the same type as what is used in the examples in the OP.

**Comment by ablaom** [comment]
@DrChainsaw Looks like your doing pretty cool things here, and would be great to get some integration at some point. We have a basic implementation of MLJ's model interface for Flux - which we are polishing at the moment. Consider it just a POC for now  (https://github.com/alan-turing-institute/MLJFlux.jl).

> If someone has a favourite search space 
 
Well, I guess this is the million dollar question. I'm no expert and would be happy your own thoughts. Do you think, if you know, that what auto-keras and tpot do is worth emulating?

**Comment by DrChainsaw** [comment]
Thanks @ablaom 

When I started I was perhaps naively thinking that the search space should not be the million dollar question if you just had a framework which allowed for arbitrary mutation of the network architecture. Turns out things might be just a little bit more complicated than that, but I still haven’t satisfied my curiosity on the subject. Once I do (or maybe before) I will certainly look into integration. From what I have seen, it looks like MLJ has answers for a lot of the API questions I have not yet wanted to address. 

I understand the neural network stuff in MLJ is in a quite early phase, but do you forsee that packages of Flux based tuning methods should depend on MLJFlux or would they depend on the more basic packages if they want to integrate with the MLJ APIs?

I’m certainly no expert either and I have not used topt or autokeras myself. I think I have seen people claim some success with them. From just browsing the code on github it was not super easy to find out what the default search space looks like. 

Reimplementing an existing and well tested method for NAS is probably a safe choice though. I’m not looking into doing this at the moment but if someone who reads this would like to tackle it I’m happy to help make use of the NaiveXX packages. If the method relies on modification of existing network architectures one should be able to save a lot of effort that by using them and I dare to say that they are quite well tested.

**Comment by ablaom** [comment]
> I understand the neural network stuff in MLJ is in a quite early phase, but do you forsee that packages of Flux based tuning methods should depend on MLJFlux or would they depend on the more basic packages if they want to integrate with the MLJ APIs?

You would depend on MLJFlux. The idea of this package is simply to encapsulate any given set of instructions for building a supervised Flux neural network  into an MLJ model (which is just a struct storing hyperparameters). In that way all the MLJ meta-algorithms (evaluation and hyperparameter optimisation) can be applied. However, competing with the flexibility implied by this remit is a desire to make deep learning models look and feel more like traditional models, to make them accessible to users outside of NLP and images. There is some friction here because the different communities value different things, but we are going to have a stab at something and see what people think.
